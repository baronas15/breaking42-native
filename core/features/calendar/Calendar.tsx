import { ScrollableTab, Tab, Tabs } from 'native-base'
import React from 'react'

import { colorTheme } from '../../shared/design/Palette.context'
import LongCalendar from './longCalendar/LongCalendar'
import WeekCalendar from './weekCalendar/WeekCalendar'

export function Calendar() {
  const styleProps = {
    activeTabStyle: { backgroundColor: colorTheme.accent },
    tabStyle: { backgroundColor: colorTheme.accent },
    textStyle: { color: '#fff' },
  }

  return (
    <>
      <Tabs
        tabBarBackgroundColor={colorTheme.accent}
        renderTabBar={() => <ScrollableTab />}>
        <Tab {...styleProps} heading="All workouts">
          <LongCalendar />
        </Tab>
        <Tab {...styleProps} heading="Weekly view">
          <WeekCalendar />
        </Tab>
      </Tabs>
    </>
  )
}
