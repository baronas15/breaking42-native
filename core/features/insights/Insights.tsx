import React, { useContext, useEffect } from 'react'
import { Text, View } from 'react-native'
import { Card } from 'react-native-elements' // DO LATER: remove this library
import { BarChart, Grid, XAxis } from 'react-native-svg-charts'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { IReducer } from '../../shared/store/Reducers.interface'
import { ApiContext } from '../../shared/utils/Api.utils'
import { ScrollContainer } from '../../universal/components/ScrollContainer'
import * as insightsActions from './Insights.actions'
import { countDistanceByDay, mapDistanceToWeekDays } from './Insights.behavior'

function Insights({ fetchPastWeekWorkouts, insights }: any) {
  const api = useContext(ApiContext)

  useEffect(() => {
    fetchPastWeekWorkouts(api)
  }, [api, fetchPastWeekWorkouts])

  // TODO: change fill
  const fill = 'rgb(10, 46, 54)'

  const contentInset = { top: 20, bottom: 20 }

  // TODO: use memo, fix any
  const data: any = mapDistanceToWeekDays(
    countDistanceByDay(insights.pastWeekData),
  )

  return (
    <ScrollContainer>
      <Card key={'weekly'} title={'Weekly training'}>
        <BarChart
          style={{ height: 200 }}
          data={data.map((x: any) => x.distance)}
          svg={{ fill }}
          contentInset={contentInset}>
          <Grid />
        </BarChart>
        <XAxis
          style={{ marginHorizontal: -10 }}
          data={data}
          formatLabel={(_: any, index: number) => data[index].day}
          contentInset={{ left: 30, right: 30 }}
          svg={{ fontSize: 10, fill: 'black' }}
        />
      </Card>

      <Card key={'Prediction'} title={'Race prediction'}>
        <View style={{ height: 150, padding: 30 }}>
          <Text>You could finish your marathon in:</Text>
          <Text>05:23:25</Text>
          <Text> </Text>
          <Text>Difference from last week:</Text>
          <Text>-04:26</Text>
        </View>
      </Card>
    </ScrollContainer>
  )
}

// DO LATER:
// Pace suggestions
// Notification to eat before run
// - 2 hour rule (wait for 2 hours after meal to start training)
// - Do your longest training runs at least three minutes per mile slower than your 5K race pace.
// - Lots of carbs on marathon week
// - New shoes after 400-500 miles (or even 300-400km)
// Graphs of performance
export default connect(
  (state: IReducer) => ({
    insights: state.insights,
  }),
  dispatch => bindActionCreators({ ...insightsActions }, dispatch),
)(Insights)
