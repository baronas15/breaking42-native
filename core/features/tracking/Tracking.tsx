import React from 'react'
import { PermissionsAndroid, Text } from 'react-native'

import { usePermission } from './effects/usePermission'
import TrackingView from './TrackingView'

const Tracking = () => {
  const permissionGranted = usePermission(
    PermissionsAndroid.PERMISSIONS.ACCESS_FINE_LOCATION,
  )

  if (!permissionGranted) {
    return <Text>Location is not enabled.</Text>
  } else {
    return <TrackingView />
  }
}

export default Tracking
