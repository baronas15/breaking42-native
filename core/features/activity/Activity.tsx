import React, { useContext, useEffect } from 'react'
import { View } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { ChangeRoute } from '../../shared/layout/routes/Router.actions'
import { ROUTE_HISTORY } from '../../shared/layout/routes/Router.constants'
import { IReducer } from '../../shared/store/Reducers.interface'
import { ApiContext } from '../../shared/utils/Api.utils'
import { Map } from '../../universal/components/Map'
import { Property } from '../../universal/components/Property'
import { SpinnerView } from '../../universal/components/SpinnerView'
import { secondsToDuration } from '../../universal/Formatting.utils'
import { useBackButton } from '../../universal/useBackButton'
import * as actions from './Activity.actions'

const Activity = ({ id, activity, fetchActivity, dispatch }: any) => {
  const api = useContext(ApiContext)

  useBackButton(() => {
    dispatch(ChangeRoute(ROUTE_HISTORY))
  })

  useEffect(() => {
    fetchActivity(api, id)
  }, [fetchActivity, api, id])

  if (!activity.gps || !activity.gps.length || !activity.general) {
    return <SpinnerView />
  }

  const TotalDistance = (
    <Property
      title={'Distance:'}
      item={`${(activity.general.distance || 0).toFixed(2).toString()} km`}
    />
  )

  return (
    <>
      <Map gps={activity.gps} />
      <View style={{ flex: 2, justifyContent: 'center', alignItems: 'center' }}>
        <Property
          title={'Time:'}
          item={secondsToDuration(activity.general.duration)}
        />
        {TotalDistance}
      </View>
    </>
  )
}

export default connect(
  (state: IReducer) => ({ activity: state.activity }),
  dispatch => bindActionCreators({ dispatch, ...actions }, dispatch),
)(Activity)
