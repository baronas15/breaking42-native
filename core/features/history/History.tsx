import { Button } from 'native-base'
import React, { useContext, useEffect } from 'react'
import { Text, View } from 'react-native'
import { Card } from 'react-native-elements' // DO LATER: remove this library
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { colorTheme } from '../../shared/design/Palette.context'
import { IReducer } from '../../shared/store/Reducers.interface'
import { ApiContext, ApiInterface } from '../../shared/utils/Api.utils'
import { IDispatch } from '../../universal/Action.interface'
import { Map } from '../../universal/components/Map'
import { Property } from '../../universal/components/Property'
import { ScrollContainer } from '../../universal/components/ScrollContainer'
import { SpinnerView } from '../../universal/components/SpinnerView'
import {
  kilometersToString,
  secondsToDuration,
  toDateString,
} from '../../universal/Formatting.utils'
import { isArrayEmpty } from '../../universal/Helper.utils'
import * as activityActions from '../activity/Activity.actions'
import * as actions from './History.actions'
import { IHistoryReducer } from './History.interface'

interface Props {
  history: IHistoryReducer
  openActivity: (activityId: string) => IDispatch
  fetchHistory: (api: ApiInterface) => IDispatch
}

const History = ({ history, fetchHistory, openActivity }: Props) => {
  const api = useContext(ApiContext)

  useEffect(() => {
    fetchHistory(api)
  }, [api, fetchHistory])

  if (isArrayEmpty(history.activities)) {
    return <SpinnerView />
  }

  return (
    <ScrollContainer footerOffset>
      <View style={{ flex: 1 }}>
        {history.activities.map(x => (
          <Card key={x._id} title={toDateString(x.general.start_time || '')}>
            <Map gps={x.gps} disableGestures />
            <View style={{ flex: 2, flexDirection: 'row' }}>
              <Property
                title={'Time:'}
                item={secondsToDuration(x.general.duration)}
              />
              <Property
                title={'Distance:'}
                item={kilometersToString(x.general.distance)}
              />
            </View>
            {/* DO LATER: Update typography and button design */}
            <Button
              block
              info
              style={{ backgroundColor: colorTheme.button }}
              onPress={() => openActivity(x._id)}>
              <Text style={{ color: '#fff' }}>VIEW NOW</Text>
            </Button>
          </Card>
        ))}
      </View>
    </ScrollContainer>
  )
}

export default connect(
  (state: IReducer) => ({
    history: state.history,
  }),
  dispatch => bindActionCreators({ ...actions, ...activityActions }, dispatch),
)(History)
