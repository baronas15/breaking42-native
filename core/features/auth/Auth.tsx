import React, { useContext } from 'react'
import { AccessToken } from 'react-native-fbsdk'
import { connect, ConnectedComponent } from 'react-redux'
import { bindActionCreators } from 'redux'

import { ChangeRoute } from '../../shared/layout/routes/Router.actions'
import { ROUTE_HISTORY } from '../../shared/layout/routes/Router.constants'
// import GoogleLogin from './GoogleButton'
import { IReducer } from '../../shared/store/Reducers.interface'
import { ApiContext } from '../../shared/utils/Api.utils'
import {
  FacebookLoginButton,
  useFacebookLogin,
} from '../../universal/components/FacebookButton'
import { SpinnerView } from '../../universal/components/SpinnerView'
import * as actions from './Auth.actions'
import { IAuthReducer } from './Auth.interface'

interface OwnProps {}

interface StateProps {
  auth: IAuthReducer
}

interface DispatchProps {
  login: Function
  dispatch: Function
}

const Auth: React.FC<StateProps & DispatchProps & OwnProps> = ({
  dispatch,
  auth,
  login,
}) => {
  const api = useContext(ApiContext)

  const [data, init, setLogIn] = useFacebookLogin((token: AccessToken) => {
    login(api, token)
  })

  if (data && !auth.loginToken) {
    login(api, data)
  } else if (auth.loginToken || data) {
    dispatch(ChangeRoute(ROUTE_HISTORY))
  }

  if (!init || data) {
    return <SpinnerView />
  }

  return <FacebookLoginButton setLogIn={(state: boolean) => setLogIn(state)} />
}

const connectedComponent: any = connect(
  (state: IReducer) => ({
    auth: state.auth,
  }),
  dispatch => bindActionCreators({ dispatch, ...actions }, dispatch),
)(Auth)

export default connectedComponent as ConnectedComponent<any, OwnProps>
