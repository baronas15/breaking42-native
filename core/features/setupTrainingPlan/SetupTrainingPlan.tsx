import { Button, Card, CardItem, Container, H1 } from 'native-base'
import React, { useContext, useState } from 'react'
import { ImageBackground, StyleSheet, Text } from 'react-native'
import { WheelPicker } from 'react-native-wheel-picker-android'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { ChangeRoute } from '../../shared/layout/routes/Router.actions'
import {
  ROUTE_CALENDAR,
  ROUTE_HISTORY,
} from '../../shared/layout/routes/Router.constants'
import { IReducer } from '../../shared/store/Reducers.interface'
import { ApiContext } from '../../shared/utils/Api.utils'
import { Form } from '../../universal/components/Form/Form'
import { FormStep } from '../../universal/components/Form/FormStep'
import { useBackButton } from '../../universal/useBackButton'
import { FirstRun } from './assets/FirstRun'
import { Performance } from './assets/Performance'
import * as actions from './SetupTrainingPlan.actions'
import { generateSchedule } from './SetupTrainingPlan.behavior'

const style = StyleSheet.create({
  container: {
    justifyContent: 'center',
    padding: 15,
    backgroundColor: 'rgba(0, 0, 0, 0.0)',
  },
  header: { alignSelf: 'center', padding: 15, color: 'white' },
  subHeader: { alignSelf: 'center', padding: 15, color: 'white' },
  spacer: { padding: 15 },
})

// DO LATER: have a back button to TrainingPlan
// DO LATER: split steps
function SetupTrainingPlan({ dispatch, setGoal }: any) {
  const api = useContext(ApiContext)

  useBackButton(() => {
    dispatch(ChangeRoute(ROUTE_HISTORY))
  })

  const mileageArray = Array.from(Array(200).keys())
  const [mileage, setMileage] = useState(0)

  return (
    <ImageBackground
      imageStyle={{ resizeMode: 'stretch' }}
      source={require('./assets/bg.jpg')}
      style={{
        flex: 1,
        justifyContent: 'center',
        width: '100%',
        height: '100%',
        opacity: 0.9,
      }}>
      <Form
        onSubmit={(data: any) => {
          dispatch(ChangeRoute(ROUTE_CALENDAR))
          const schedule = generateSchedule(data)
          const result = { ...data, schedule }
          setGoal(api, result)
        }}
        initial={{ step: 'choseGoal' }}
        render={({ updateForm, next, submit }: any) => (
          <>
            <FormStep identifier="choseGoal">
              <Container style={style.container}>
                <H1 style={style.header}>What is your goal?</H1>
                <Text style={style.subHeader}>
                  It seems that you don't have a training plan yet, but first we
                  need some information about you
                </Text>
                <Card>
                  <CardItem
                    button
                    onPress={() => {
                      updateForm({
                        key: 'target.goalType',
                        required: true,
                        value: 'first',
                        nextStep: 'targetDistance',
                      })
                    }}>
                    <FirstRun />
                    <Text style={style.spacer}>
                      First race (5K, 10K, half marathon, marathon)
                    </Text>
                  </CardItem>
                </Card>
                <Card>
                  <CardItem
                    button
                    onPress={() => {
                      updateForm({
                        key: 'target.goalType',
                        required: true,
                        value: 'performance',
                        nextStep: 'targetDistance',
                      })
                    }}>
                    <Performance />
                    <Text style={style.spacer}>Improve performance</Text>
                  </CardItem>
                </Card>
              </Container>
            </FormStep>
            <FormStep identifier="targetDistance">
              <Container style={style.container}>
                <H1 style={style.header}>What is your target distance?</H1>
                {[
                  { key: '5k', text: '5k' },
                  { key: '10k', text: '10k' },
                  { key: 'halfMarathon', text: 'Half marathon' },
                  { key: 'marathon', text: 'Marathon' },
                ].map(x => (
                  <Card>
                    <CardItem
                      button
                      onPress={() => {
                        updateForm({
                          key: 'target.distance',
                          required: true,
                          value: x.key,
                          nextStep: 'recentBest',
                        })
                      }}>
                      <Text>{x.text}</Text>
                    </CardItem>
                  </Card>
                ))}
              </Container>
            </FormStep>
            <FormStep identifier="recentBest">
              <Container style={style.container}>
                <H1 style={style.header}>What was your mileage past week?</H1>
                <Card>
                  <CardItem style={{ justifyContent: 'center' }}>
                    <WheelPicker
                      itemTextFontFamily={''}
                      selectedItemTextFontFamily={''}
                      data={mileageArray.map(value => `${value} km`)}
                      selectedItem={mileage}
                      onItemSelected={(index: number) => {
                        updateForm({
                          key: 'startingMileage',
                          required: true,
                          value: mileageArray[index],
                        })

                        setMileage(index)
                      }}
                      style={{ width: 150, height: 130 }}
                    />
                  </CardItem>
                  <Button
                    block
                    onPress={() => {
                      next('workoutCount')
                    }}>
                    <Text style={{ color: 'white' }}>Next</Text>
                  </Button>
                </Card>
              </Container>
            </FormStep>
            <FormStep identifier="workoutCount">
              <Container
                style={{ ...style.container, justifyContent: 'center' }}>
                <H1 style={style.header}>
                  How many workouts a week can you have?
                </H1>
                <Card>
                  {[
                    { key: '3', text: '3 workouts' },
                    { key: '4', text: '4 workouts' },
                    { key: '5', text: '5 workouts' },
                    { key: '6', text: '6 workouts' },
                  ].map(x => (
                    <CardItem
                      button
                      onPress={() => {
                        updateForm({
                          key: 'weeklyWorkouts',
                          required: true,
                          numeric: true,
                          value: x.key,
                          nextStep: 'submit',
                        })
                      }}>
                      <Text>{x.text}</Text>
                    </CardItem>
                  ))}
                </Card>
              </Container>
            </FormStep>
            <FormStep identifier="submit">
              <Container
                style={{ ...style.container, justifyContent: 'center' }}>
                <Button
                  block
                  info
                  onPress={() => {
                    submit()
                  }}>
                  <Text style={{ color: 'white' }}>Set a goal</Text>
                </Button>
                <Button
                  block
                  light
                  onPress={() => dispatch(ChangeRoute(ROUTE_HISTORY))}>
                  <Text style={{ color: 'white' }}>Cancel</Text>
                </Button>
              </Container>
            </FormStep>
          </>
        )}
      />
    </ImageBackground>
  )
}

export default connect(
  (_: IReducer) => ({}),
  dispatch => bindActionCreators({ dispatch, ...actions }, dispatch),
)(SetupTrainingPlan)
