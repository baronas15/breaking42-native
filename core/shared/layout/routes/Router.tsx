import React from 'react'
import { StatusBar } from 'react-native'
import { connect } from 'react-redux'

import Activity from '../../../features/activity/Activity'
import Auth from '../../../features/auth/Auth'
import { IAuthReducer } from '../../../features/auth/Auth.interface'
import { Calendar } from '../../../features/calendar/Calendar'
import History from '../../../features/history/History'
import Insights from '../../../features/insights/Insights'
import MoreSettings from '../../../features/moreSettings/MoreSettings'
import SelectWorkout from '../../../features/selectWorkout/SelectWorkout'
import SetupTrainingPlan from '../../../features/setupTrainingPlan/SetupTrainingPlan'
import Tracking from '../../../features/tracking/Tracking'
import TrainingPlan from '../../../features/trainingPlan/TrainingPlan'
import { colorTheme } from '../../design/Palette.context'
import { IReducer } from '../../store/Reducers.interface'
import {
  ApiContext,
  loginApi,
  queryApi,
  workoutsApi,
} from '../../utils/Api.utils'
import MainLayout from '../MainLayout'
import * as actions from './Router.actions'
import {
  ROUTE_ACTIVITY_RESULT,
  ROUTE_CALENDAR,
  ROUTE_HISTORY,
  ROUTE_INSIGHTS,
  ROUTE_MORE_SETTINGS,
  ROUTE_NONE,
  ROUTE_SELECT_WORKOUT,
  ROUTE_SETUP_TRAINING_PLAN,
  ROUTE_TRACKING,
  ROUTE_TRAINING_PLAN,
} from './Router.constants'
import { IRouterReducer } from './Router.interface'

interface Props {
  router: IRouterReducer
  auth: IAuthReducer
}

const withLayout = (Component: any, props?: any) => {
  return (
    <MainLayout>
      <Component {...props} />
    </MainLayout>
  )
}

function Router(props: Props) {
  if (props.router.route === ROUTE_NONE || !props.auth.loginToken) {
    return <Auth />
  }

  return (
    <ApiContext.Provider
      value={{
        loginApi: loginApi(''),
        queryApi: queryApi(props.auth.loginToken || ''),
        workoutsApi: workoutsApi(props.auth.loginToken || ''),
      }}>
      <StatusBar backgroundColor={colorTheme.accent} />

      <PickRoute {...props} />
    </ApiContext.Provider>
  )
}

function PickRoute(props: Props) {
  switch (props.router.route) {
    case ROUTE_TRACKING:
      return withLayout(Tracking)
    case ROUTE_INSIGHTS:
      return withLayout(Insights)
    case ROUTE_ACTIVITY_RESULT:
      return withLayout(Activity, {
        id: props.router.params ? props.router.params.id : '',
      })
    case ROUTE_CALENDAR:
      return withLayout(Calendar)
    case ROUTE_MORE_SETTINGS:
      return withLayout(MoreSettings)
    case ROUTE_TRAINING_PLAN:
      return withLayout(TrainingPlan)
    case ROUTE_SELECT_WORKOUT:
      return <SelectWorkout />
    case ROUTE_SETUP_TRAINING_PLAN:
      return <SetupTrainingPlan />
    case ROUTE_HISTORY:
    default:
      return withLayout(History)
  }
}

export default connect(
  (state: IReducer) => ({
    router: state.router,
    auth: state.auth,
  }),
  actions,
)(Router)
