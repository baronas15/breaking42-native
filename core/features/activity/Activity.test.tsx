import React from 'react'
import { Provider } from 'react-redux'
import renderer from 'react-test-renderer'
import configureStore from 'redux-mock-store'

import { defaultRouterState } from '../../shared/layout/routes/Router.reducer'
import { IReducer } from '../../shared/store/Reducers.interface'
import { defaultAuthState } from '../auth/Auth.reducer'
import { defaultCalendarState } from '../calendar/Calendar.reducer'
import { defaultHistoryState } from '../history/History.reducer'
import { defaultInsightsState } from '../insights/Insights.reducer'
import { defaultGoalState } from '../setupTrainingPlan/SetupTrainingPlan.reducer'
import { defaultTrackingState } from '../tracking/Tracking.reducer'
import Activity from './Activity'

const mockStore = configureStore([])

describe('Activity', () => {
  describe('component', () => {
    let store: any
    let component: any

    beforeEach(() => {
      const state: IReducer = {
        activity: {
          _id: '1',
          userId: '1',
          general: {
            calories: 500,
            distance: 12,
            duration: 3600,
            hydration: 0,
            start_time: '2020-05-05',
          },
          gps: [],
          laps: [],
        },
        auth: defaultAuthState,
        calendar: defaultCalendarState,
        goal: defaultGoalState,
        history: defaultHistoryState,
        insights: defaultInsightsState,
        router: defaultRouterState,
        tracking: defaultTrackingState,
      }
      store = mockStore(state)

      const ActivityComponent: any = Activity

      component = renderer.create(
        <Provider store={store}>
          <ActivityComponent id={'1'} />
        </Provider>,
      )
    })

    it('should generate component', () => {
      let tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
