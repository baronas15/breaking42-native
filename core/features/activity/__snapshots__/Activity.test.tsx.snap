// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity component should generate component 1`] = `
<View
  style={
    Object {
      "alignContent": "center",
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#45D56E"
    hidesWhenStopped={true}
    size="large"
    style={
      Object {
        "height": 80,
      }
    }
  />
</View>
`;
